pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,215 @@
+				Provider: &pluginProvider{
+					matchURL: imageMatchRef.MatchURL,
+					plugin: &execPlugin{
+						name:         plugin.Name,
+						pluginBinDir: pluginBinDir,
+						args:         convertArgs(plugin.ExtraArgs),
+					},
+				},
+				Lifetime: imageMatchRef.CacheDuration.Duration,
+			}
+
+			// The registered provider name is the plugin name + the match URL of the image reference since
+			// each image match reference in a given plugin is registered as a separate credential provider
+			// and each registered name must be unique.
+			providerName := plugin.Name + "-" + imageMatchRef.MatchURL
+			credentialprovider.RegisterCredentialProvider(providerName, cachingProvider)
+		}
+	}
+
+	return nil
+}
+
+// pluginProvider is the plugin-based implementation of the DockerConfigProvider interface.
+type pluginProvider struct {
+	matchURL string
+
+	plugin Plugin
+}
+
+// Provide returns a credentialprovider.DockerConfig based on the credentials returned
+// by the exec plugin.
+func (p *pluginProvider) Provide(image string) credentialprovider.DockerConfig {
+	cfg := make(credentialprovider.DockerConfig)
+
+	if matched, _ := credentialprovider.URLsMatchStr(p.matchURL, image); !matched {
+		return cfg
+	}
+
+	response, err := p.plugin.ExecPlugin(image)
+	if err != nil {
+		klog.Errorf("Failed getting credential from external registry credential provider: %v", err)
+		return cfg
+	}
+
+	cfg[p.matchURL] = credentialprovider.DockerConfigEntry{
+		Username: response.Username,
+		Password: response.Password,
+		Email:    "plugin@criauth.k8s.io", // email is deprecated, this is just a dummy email
+	}
+	return cfg
+}
+
+// Enabled always returns true since registration of the plugin via kubelet implies it should be enabled.
+func (e *pluginProvider) Enabled() bool {
+	return true
+}
+
+// IsPlugin returns true since all external credential providers are plugins
+func (e *pluginProvider) IsPlugin() bool {
+	return true
+}
+
+// Plugin is the interface calling ExecPlugin. This is mainly for testability
+// so tests don't have to actually exec any processes.
+type Plugin interface {
+	ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error)
+}
+
+// execPlugin is the implementation of the Plugin interface that execs a provided CRI auth plugin based
+// on it's name provided in CRIAuthPluginConfig. It is assumed that the executable is available in the
+// plugin directory provided by the kubelet.
+type execPlugin struct {
+	name         string
+	args         []string
+	pluginBinDir string
+}
+
+// ExecPlugin executes the plugin binary in the following format:
+//
+//  $ <plugin-name> get-credentials arg1=value1 arg2=value2
+//
+// The plugin is expected to receive the CRIAuthPluginRequest API via stdin and is expected
+// to return the CRIAuthPluginResponse via stdout.
+func (e *execPlugin) ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error) {
+	pluginBin := e.pluginBinDir + "/" + e.name
```

Comment:
nit: I understand the configuration of plugins would be ACL'd. So this is not a big deal. But in general, when constructing path like this - it is a good practice to make sure that `e.name` doesn't start with `../`

Reply:
Hmm, good call. So the ordering is only guarenteed between external plugins and the built-in ones here: https://github.com/kubernetes/kubernetes/pull/94196/files#diff-8bcfde5fa58ec8d8ae4b306dd85768a2R53-R84 

Order only amongst the external plugins and the builtin plugins is not guarenteeed.